File: mathematics/sum-of-four-squares/sum-of-four-squares.py
import math
from typing import List, Optional


def find_perfect_square(num: int) -> Optional[int]:
    sqrt = math.isqrt(num)
    return sqrt if sqrt * sqrt == num else None


def find_three_squares(num: int) -> Optional[List[int]]:
    m = num
    while m % 4 == 0:
        m //= 4
    if m % 8 == 7:
        return None

    for a in range(int(math.isqrt(num)) + 1):
        for b in range(int(math.isqrt(num - a * a)) + 1):
            c2 = num - a * a - b * b
            if c2 >= 0:
                c = math.isqrt(c2)
                if c * c == c2:
                    return [a, b, c]
    return None


def sum_of_four_squares(input: List[int]):
    num_of_test_cases = input[0]
    n_values = input[1:]

    if num_of_test_cases < 1 or num_of_test_cases > 1000:
        print("The number of test cases must be at least 1 and at most 1000")
        return

    if num_of_test_cases != len(n_values):
        print("The num_of_test_cases must equal the available test cases")

    for n in n_values:
        if n == 0:
            print("0 0 0 0")
            continue

        sqrt = find_perfect_square(n)
        if sqrt is not None:
            print(f"{sqrt} 0 0 0")
            continue

        for a in range(int(math.isqrt(n)), -1, -1):
            r = n - a * a
            three_squares = find_three_squares(r)
            if three_squares:
                print(f"{three_squares[0]} {three_squares[1]} {three_squares[2]} {a}")
                break


sum_of_four_squares([5, 5, 30, 74, 17, 322266])

File: mathematics/sum-of-four-squares/sum-of-four-squares.ts
function sumOfFourSquares(input: Array<number>) {
  const numTestCases = input[0];
  const nValues = input.slice(1);

  if (numTestCases < 1 || numTestCases > 1000) {
    console.log("The number of test cases must be at least 1 and at most 1000");
    return;
  }

  if (numTestCases !== nValues.length) {
    console.log("The num_of_test_cases must equal the available test cases");
    return;
  }

  for (const n of nValues) {
    if (n === 0) {
      console.log(`0 0 0 0`);
      continue;
    }

    const sqrt = findPerfectSquare(n);
    if (sqrt !== false) {
      console.log(`${sqrt} 0 0 0`);
      continue;
    }

    let found = false;
    for (let a = Math.floor(Math.sqrt(n)); a >= 0 && !found; a--) {
      const r = n - a * a;
      const threeSquares = findThreeSquares(r);
      if (threeSquares) {
        console.log(
          `${threeSquares[0]} ${threeSquares[1]} ${threeSquares[2]} ${a}`
        );
        found = true;
      }
    }
  }
}

function findPerfectSquare(num: number): number | false {
  const sqrt = Math.floor(Math.sqrt(num));
  return sqrt * sqrt === num ? sqrt : false;
}

function findThreeSquares(num: number): Array<number> | false {
  // Legendre's three-square theorem
  let m = num;
  while (m % 4 === 0) m /= 4;
  if (m % 8 === 7) return false;

  for (let a = 0; a * a <= num; a++) {
    for (let b = 0; b * b <= num - a * a; b++) {
      const c2 = num - a * a - b * b;
      if (c2 >= 0) {
        const c = Math.floor(Math.sqrt(c2));
        if (c * c === c2) {
          return [a, b, c];
        }
      }
    }
  }
  return false;
}

sumOfFourSquares([5, 5, 30, 74, 17, 322266]);
